[CONT #32683] os_env_defined() and os_getenv_noalloc()

juditnovak opened the pull request:
This PR is improving performance in terms of memory allocation on top of changes in [ [BUGFIX] Removing internal 'envmap' from environment evaluation #32683 ](https://github.com/neovim/neovim/pull/32683).

The two enhancements that were suggested there, implemented here:
 - `os_env_defined()` (See [thread](https://github.com/neovim/neovim/pull/32683#pullrequestreview-2676706533)
 - `os_env_noalloc()` (See [comment](https://github.com/neovim/neovim/pull/32683#issuecomment-2711951176))

juditnovak commented:
Note: [pipeline](https://github.com/juditnovak/neovim/actions/runs/13861169266/job/38789780935) :green_circle:

juditnovak commented on src/nvim/os/stdpaths.c:
```diff
@@ -71,20 +71,19 @@ static const char *const xdg_defaults[] = {
 const char *get_appname(bool namelike)
 {
-  const char *env_val = os_getenv("NVIM_APPNAME");
+  const char *env_val = os_getenv_noalloc("NVIM_APPNAME");
```
THIS CHANGE SHOULD NOT BE MERGED, IT'S A **BREAKING CHANGE***,

I only left this code for the sake of demonstration, with the tweaked `env_spec.lua`.

juditnovak commented on test/functional/options/defaults_spec.lua:
```diff
@@ -922,7 +922,7 @@ describe('stdpath()', function()
 
   it('$NVIM_APPNAME', function()
-    local appname = 'NVIM_APPNAME_TEST' .. ('_'):rep(106)
+    local appname = 'NVIM_APPNAME_TEST' .. ('_'):rep(46)
```
Demonstrating **BREAKING CHANGE**, mustn't be merged

juditnovak commented:
@justinmk , @zeertzjq you are welcome to take a look at changes following your suggestions on [ [BUGFIX] Removing internal 'envmap' from environment evaluation #32683 ](https://github.com/neovim/neovim/pull/32683).

(Given the complexity, I preferred to split functionally complete chapters of this work in two stages. Once agreed, of course I'm glad to add the contents of this branch to the original PR.)

zeertzjq commented on src/nvim/os/os_win_console.c:
```diff
@@ -118,9 +118,7 @@ void os_title_reset(void)
-  const char *conemuansi_env = os_getenv("ConEmuANSI");
-  bool conemu_ansi = strequal(conemuansi_env, "ON");
-  xfree((char *)conemuansi_env);
+  bool conemu_ansi = strequal(os_getenv("ConEmuANSI"), "ON");
```
Missing `_noalloc`?

juditnovak replied:
:facepalm:  Thank you so much, very good call!!!!

juditnovak replied:
Done.

zeertzjq commented on src/nvim/os/lang.c:
```diff
@@ -70,7 +70,6 @@ char *get_mess_lang(void)
 }
 
 /// Get the language used for messages from the environment.
-/// Result must be freed by the caller.
```
Specify that this function may use `NameBuff`?

juditnovak replied:
Done.

justinmk commented on src/nvim/os/env.c:
```diff
@@ -109,6 +129,26 @@ bool os_env_exists(const char *name)
 
+/// Returns true if environment variable `name` is defined (i.e. exists and not empty).
+/// Returns false if not found (UV_ENOENT) or other failure.
+bool os_env_defined(const char *name)
```
can we just make this a param on `os_env_exists`

juditnovak replied:
@zeertzjq would you agree to that?

(I implemented [his original suggestion](https://github.com/neovim/neovim/pull/32683#pullrequestreview-2676706533), but I'm good to turn it into a param on `os_env_exists()`)

zeertzjq replied:
Yes

juditnovak replied:
Done.

justinmk commented:
> I preferred to split functionally complete chapters of this work in two stages. Once agreed, of course I'm glad to add the contents of this branch to the original PR.)

thanks, but please create the PRs in the core repo. Else the discussion here won't be found in searches, in the future.

juditnovak commented:
@justinmk I see your point. I've addressed the comments from here, and added all changes from here back to the original PR (which is against the `neovim/neovim` repository). We can continue the disucssion there.

justinmk commented:
If you want to create separate incremental PRs in the neovim repo, that's fine too.
